New blog about how to create and publish packages (#295)
lf-lang/lf-lang.github.io: blog/2024-11-11-package.md

{/* truncate */}

In this guide, we’ll walk you through the essential steps to create and publish a package to the community repository, covering:

1. [**Creating a New Package**](#creating-a-package)  
   Get started by setting up your package with the required files and directory structure to meet Lingua Franca standards.

2. [**Configuring the `Lingo.toml` File**](#configuring-the-lingotoml-file)  
   Set up the `Lingo.toml` configuration to ensure compatibility with the Lingo Package Manager, making your package easy to download and install.

3. [**Publishing to your (public) repository**](#publishing-a-package)  
   Learn how to publish your package to share it with the broader community.

By the end of this guide, you'll have a fully configured, shareable package ready to distribute within the Lingua Franca ecosystem.

## Creating a New Package

> You can create a new [LF package](/docs/glossary/#package) either manually by creating an [LF file](/docs/glossary/#lf-file) or by using the [Lingo Package Manager](https://github.com/lf-lang/lingo).

Reviewer: ```suggestion
You can create a new [LF package](/docs/glossary/#package) using the [Lingo Package Manager](https://github.com/lf-lang/lingo) (recommended) or manually by creating the required files.
```

Author: You're suggesting we only describe Option 1: "Create a Project Using the Lingo Package Manager" and remove Option 2. While this would streamline the text, I would still include a note for those creating the project manually, highlighting where to place the `Lingo.toml` file.

> If you manually create the `Lingo.toml` file, place it adjacent to the `src` folder in the root directory of the package. This file serves as a configuration for the package, allowing you to specify the package name, version, and other metadata, including any dependencies you want to install.

Reviewer: I wasn’t suggesting omitting option 2, at least not intentionally. ??

#### Option 1: Create a Project Using the Lingo Package Manager (Recommended)
1. After [installing the Lingo Package Manager](https://www.lf-lang.org/docs/installation#lingo), create an empty directory to serve as the root of your new package.
2. Open the terminal in this folder and run the <kbd>lingo init</kbd> command. 

This will set up a new LF package with the following structure:

```
├── .
│   ├── src/
│   │  └── Main.lf
└── └── Lingo.toml # Configuration file for current package
```

> #### Option 2: Create a New LF File

Reviewer: ```suggestion
#### Option 2: Manually Create the Required Files
```

...
```

In LF, there is no visibility control, meaning you cannot mark a reactor or file as `private` as in other programming languages. The `private` directory serves this purpose by containing files meant solely for internal package use. Files in this directory are hidden in the Package Explorer within the VS Code extension, making them inaccessible to other packages. The `private` directory is useful for implementing design patterns like the Abstract Factory Pattern, which provides an interface for creating families of related objects without specifying concrete classes. For example, a developer might define an `AbstractLibraryFile.lf` containing abstract reactors and implement them in various ways within different files under the `lib` directory.

### Must-Follow Guidelines

When creating a new package, ensure that you follow these guidelines:

- **README.md**: Include a `README.md` file in the root directory. This should provide an overview of the package, its purpose, a description of the LF files in the `lib/` folder, and relevant user information. The `README.md` is essential for publishing your package to the community repository and must be kept up to date.

> 

Reviewer: ```suggestion

- **LICENSE**: A text file giving the license terms for the package.
```

- **Example LF Files**: Include example LF files in the `src/` directory to demonstrate how to use the reactors in the `lib/` directory. These examples should be well-documented with comments explaining the purpose and usage of each reactor.

## Configuring the `Lingo.toml` File

The `Lingo.toml` file is a configuration file that provides essential information about your package, such as its name, version, and dependencies. It ensures that your package is recognized and installed correctly by the Lingo Package Manager. The `Lingo.toml` file should include the following sections:

- **`[package]`**: Specify the package name, description, and version.

- **`[lib]`**: Define the package name, the target language used to implement the code, the main reactor (which could be one of the examples from the `src/` directory), and the platform used to build the package.

Here's an example of a `Lingo.toml` file:

```toml
[package]
name = "PackageName"
version = "0.3.1"

[properties]

[lib]
name = "PackageName"

> main = "./src/Main.lf"

Reviewer: It is not clear to me what `main` should be for a library. Typically, a library will have multiple example programs illustrating the use of the library.  What file should this point to?

Author: In my case, I simply added one of the several `.lf` programs I had developed as an example. I assume there isn’t a single definitive `main` program when dealing with libraries.

Reviewer: @tanneberger tells me that a library should have a location entry rather than main.  Unfortunately, I don’t think this is documented anywhere, so I guess this blog will become he primary documentation.  Maybe you two need to get together to make this clear?

{/* truncate */}

In this guide, we’ll walk you through the essential steps to create and publish a package to the community repository, covering:

1. [**Creating a New Package**](#creating-a-package)  
   Get started by setting up your package with the required files and directory structure to meet Lingua Franca standards.

2. [**Configuring the `Lingo.toml` File**](#configuring-the-lingotoml-file)  
   Set up the `Lingo.toml` configuration to ensure compatibility with the Lingo Package Manager, making your package easy to download and install.

3. [**Publishing to the Community Repository**](#publishing-to-the-community-repository)  
   Learn how to publish your package to the [Lingua Franca Community Package Repository](https://github.com/lf-pkgs) to share it with the broader community.

By the end of this guide, you'll have a fully configured, shareable package ready to distribute within the Lingua Franca ecosystem.

## Creating a New Package

> You can create a new [LF package](/docs/glossary/#package) either manually by creating an [LF file](/docs/glossary/#lf-file) or by using the [Lingo Package Manager](https://github.com/lf-lang/lingo).

Author: Yes, I can add `(recommended)` next to Option 1, which involves using Lingo.

> And a single LF file does not classify as a package.

That’s correct. In fact, there is already a link to the Glossary explaining what an LF file is. However, it serves as the starting point for creating a package, which should include additional elements (there is a definition for LF package too in the Glossary)

#### Option 1: Create a Project Using the Lingo Package Manager
1. After [installing the Lingo Package Manager](https://www.lf-lang.org/docs/installation#lingo), create an empty directory to serve as the root of your new package.

> 2. Open the folder in VS Code.

Author: What should I add about `--platform` and `--language` here?

> 2. Open the folder in VS Code.

Reviewer: I'm confused about this. `--platform` and `--language` are not documented anywhere that I could find. They are not accepted as command-line options to `lingo`.

3. Open the terminal in this folder and run the <kbd>lingo init</kbd> command. 

This will set up a new LF package with the following structure:

```
├── .
│   ├── src/
│   │  └── Main.lf
└── └── Lingo.toml # Configuration file for current package
```

> #### Option 2: Create a New LF File

Author: But this is the structure that I get when I execute `lingo init`

> #### Option 2: Create a New LF File

Reviewer: I have verified that this is exactly what `lingo init` does.

## Configuring the `Lingo.toml` File

The `Lingo.toml` file is a configuration file that provides essential information about your package, such as its name, version, and dependencies. It ensures that your package is recognized and installed correctly by the Lingo Package Manager. The `Lingo.toml` file should include the following sections:

- **`[package]`**: Specify the package name, description, and version.

- **`[lib]`**: Define the package name, the target language used to implement the code, the main reactor (which could be one of the examples from the `src/` directory), and the platform used to build the package.

Here's an example of a `Lingo.toml` file:

```toml
[package]
name = "PackageName"
version = "0.3.1"

[properties]

[lib]
name = "PackageName"
main = "./src/Main.lf"
target = "Python"
platform = "Native"

[lib.properties]

[dependencies]
```

The `Lingo.toml` can also include additional sections for dependencies and other purpose-specific configurations, however, these are optional and can be added as needed.

## Publishing to the Community Repository

> Once you've created your package and configured the `Lingo.toml` file, you’re ready to publish it in the [Lingua Franca Packages](https://github.com/lf-pkgs) organization. Publishing your package here allows other developers to easily find, install, and use it in their projects.

Reviewer 2: I think the default workflow should be the folks publish to their own (public) repo. That will already allow for the exchange of reusable libraries. Then, later on in the description we could mention the `lf-pkgs` organization and getting people's repos transfered. I'm actually starting to doubt that this is the right approach at all. I think it might be better to have a `lf-lang/pkgs` repo with a single text file that lists all the repos. That would be super easy to build a little website around where people can search for packages (or a feature in VS Code, for that matter). In the near term, adding a package would just amount to adding the repo URL, and it would not require transferring ownership. Ownership transfer is tedious because anyone who transfers ownership would have to be a member of the `lf-lang` organization, which has paid seats, so this won't scale. Perhaps this blog should just omit these details? Also tagging @tanneberger for feedback on this.

Author: > I think it might be better to have a lf-lang/pkgs repo with a single text file that lists all the repos.

I agree with that, it will be easier to show the list of packages, and people in the community won’t need to ask to join the GitHub organization.

So, in this blog, I’ll leave out all the comments about the GitHub organization for now

Reviewer 2: On reflection, if we do that route, we don't even need a separate organization. We could just have a `lf-lang/pkgs` repo.

### Must-follow guidelines

When creating a new package, ensure that you follow these guidelines:

- **README.md**: Include a `README.md` file in the root directory. This should provide an overview of the package, its purpose, a description of the LF files in the `lib/` folder, and relevant user information. The `README.md` is essential for publishing your package to the community repository and must be kept up to date.

- **LICENSE**: A text file giving the license terms for the package.
- **File Naming**: Use descriptive names for files in the `lib/` directory that reflect their purpose. For example, name a file containing a reactor for different sensors as `Sensor.lf`.

- **Example LF Files**: Include example LF files in the `src/` directory to demonstrate how to use the reactors in the `lib/` directory. These examples should be well-documented with comments explaining the purpose and usage of each reactor.

> ## Configuring the `Lingo.toml` File

Author: @tanneberger I think we need to coordinate to provide more info about `Lingo.toml`
